### ISSUE TITLE
[stable/wordpress] can't use https behind ingress

### ISSUE THREAD

**Issue by kachkaev** [helm/charts#1567]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

bug report

**Version of Helm and Kubernetes**:

2.5.0 / 1.6.7 (should not be relevant)

**Which chart**:

stable/wordpress

**What happened**:

Looks like the chart is not compatible with HTTPS termination by ingress. When HTTP is automatically redirected to HTTPS and then the website is loaded, all the the wordpress assets and external links still point to `http://example.com` instead of `https://example.com`. Unfortunately, this makes the deployed website completely unusable even after enabling insecure content in the browser ![insecure](https://user-images.githubusercontent.com/608862/28713538-52d3596a-7387-11e7-9c20-b7478601f3a0.png). The biggest problem is that the forms point to HTTP, so attempts to login or post a comment are bounced by 301 from ingress so no POST data reaches the pod.

![bitnami-wordpress-https](https://user-images.githubusercontent.com/608862/28713493-257a052c-7387-11e7-83db-4f49db9b66f1.png)

**What you expected to happen**:

I expected to see a working wordpress website when ingress is enabled and HTTPS is enforced.

**How to reproduce it** (as minimally and precisely as possible):

1.  Make sure that ingress is configured to redirect HTTP to HTTPS for all its rules (e.g. when in [traefik](https://github.com/kubernetes/charts/tree/master/stable/traefik)'s `ssl` config section `ssl.enabled` && `ssl.enforced` are `true`).

2.  Install the latest stable wordpress chart with 
```bash
ingress:
    enabled: true
    hostname: example.com
```

**Anything else we need to know**:

The issue is also reported upstream https://github.com/bitnami/bitnami-docker-wordpress/issues/101

**Comment by sameersbn** [comment]
Have you tried to force https by installing https://wordpress.org/plugins/wp-force-https/
Please note the pr https://github.com/kubernetes/charts/pull/1475 would need to be merged for this to work properly. Could you please test it and let us know if it works?

**Comment by keunwoo** [comment]
@sameersbn 

> Have you tried to force https by installing https://wordpress.org/plugins/wp-force-https/

The "Force HTTPS" plugin doesn't work, because it serves a redirect when the inbound connection to WordPress itself is http, and the ingress controller will be speaking http to WordPress. The result of installing that plugin is an infinite redirect on all routes from the https URL to itself.

The documented hack for WordPress behind an SSL-terminating reverse proxy is to check for the header that the proxy applies. For nginx, the header is `X-Forwarded-Proto`, and the following recipe is recommended:

https://codex.wordpress.org/Administration_Over_SSL#Using_a_Reverse_Proxy
https://feliciano.tech/blog/running-wordpress-behind-an-sslhttps-terminating-proxy/

It is mildly surprising that the recommended deployment method is hand-editing a branch into the the wp-config.php file. Some further searching reveals that the ["Insecure Content Fixer" plugin](https://wordpress.org/plugins/ssl-insecure-content-fixer/) has a built-in way of configuring a check for this header:

https://blog.virtualzone.de/2016/08/howto-https-ssl-wordpress-behind-proxy-nginx-haproxy-apache-lighttpd.html

Adapting this recipe, I seem to have gotten things working via the following steps:

1. Deploy the chart with a non-TLS ingress. Your deployment configuration should explicitly specify the `wordpressPassword` property, for reasons described below [0].
2. Login as admin and install the "Insecure Content Fixer" plugin. Configure that plugin to use HTTP_X_FORWARDED_PROTO detection, or whatever method is appropriate for your ingress controller.
3. Redeploy the chart with a TLS-enabled ingress using `helm update`.
4. Visit the https version of your site and verify that everything is working.

Note that virtualzone.de post above suggests configuring the "WordPress Address (URL)" and "Site Address (URL)" settings to reflect the https URL. These fields are disabled (present, but not editable) under Bitnami-packaged WordPress; my guess is that they are [hardcoded into the wp-config.php file](https://codex.wordpress.org/Changing_The_Site_URL#Edit_wp-config.php), although I have not ssh'd onto the node to verify this. However, the procedure I describe above worked anyway. Since I don't know WordPress well, I'm not sure why. Regardless, it seems like Bitnami's WordPress packaging should permit some method of configuring these fields at deployment time.

Anyway, the 4-step point-and-click fiddling that I had to do above seems like exactly what helm was designed to obviate. Thus, unless I'm misunderstanding something, it seems like running behind a reverse proxy (i.e. with a TLS-enabled ingress controller) should not yet be considered a fully supported mode of operation for this chart. Maybe someone who was involved in the implementation can describe whether it was ever observed to work, and under what configuration?

[0] The reason that you should specify an explicit WordPress admin password in your deployment configuration is that apparently if you rely on password auto-generation, `helm update` will generate a new password and stash it in the Kubernetes secret, but fail to update the WordPress installation to reflect this change. Therefore the password you obtain by `kubectl get secret ... -o jsonpath="{.data.wordpress-password}"` will be incorrect after a `helm update`.

**Comment by kachkaev** [comment]
Thanks for sharing your thoughts @keunwoo! I repeated the steps and the workaround was successful! Just to avoid any confusion by others, the exact name of the wordpress plugin to install is [_SSL_ Insecure Content Fixer](https://wordpress.org/plugins/ssl-insecure-content-fixer/).

Totally agree with you that the process of making wordpress HTTPS-compatible is somewhat less straightforward than one would expect from a stable k8s chart. I wish I knew wordpress enough to suggest a fix in the bitnami image or the chart itself, but I'm afraid that's beyond what I'll be able to offer.

@sameersbn @chipironcin what would you suggest as a proper fix? Although I don't have any stats in front of me, I suspect that running wordpress behind an HTTPS-enabled ingress may be quite a popular thing. It is therefore important to do something with this issue until too many people stumbled across it.

**Comment by prydonius** [comment]
After talking about this with @chipironcin I think we need both a WORDPRESS_HOST and WORDPRESS_FORCE_TLS parameter. By default, WordPress is configured to use the current host as the site url (HTTP_HOST header), but the config currently hardcodes the scheme of this to HTTP. WORDPRESS_FORCE_TLS could keep this the same but hardcode HTTPS instead of HTTP.

Alternatively, if you know the URL, we could support a WORDPRESS_HOST env var that gets used instead of HTTP_HOST.

**Comment by kachkaev** [comment]
Thanks @prydonius. I might be wrong, but I think we should be a bit careful with assigning `WORDPRESS_HOST` or `WORDPRESS_FORCE_TLS` environment variables to the container. I can imagine a situation when someone decides to move their website from one domain to another or toggle HTTPS – the pod will need to be destroyed and recreated.

The host and the protocol do reach the container via HTTP headers, so in theory this information is enough. Because `WORDPRESS_HOST` is absent in [deployment.yaml](https://github.com/kubernetes/charts/blob/0d485610e6af1ba121898db27a32c50143e39636/stable/wordpress/templates/deployment.yaml), I guess the domain detection already works well. In the ideal scenario, we just need to teach the bitnami wordpress image to detect the protocol as well and that will be it. After looking at the the current port, wordpress could check `X-Forwarded-Proto` and override the protocol if this value is present. Can there be a wordpress plugin for this that can be installed by default?

In addition, it'd be useful to consider `X-Forwarded-For` header, because it contains the correct remote IP for the request that go through the ingress controller. The same strategy can be applied: we first check `$_SERVER['REMOTE_ADDR']` (or whatever) and if `X-Forwarded-For` header is present, value is overridden from there.

@chipironcin what do you think of updating [bitnami-docker-wordpress](https://github.com/bitnami/bitnami-docker-wordpress) to cover both cases? We'll then just need to update the image tag in the helm chart. Bitnami users outside helm will also benefit.

**Comment by prydonius** [comment]
> Thanks @prydonius. I might be wrong, but I think we should be a bit careful with assigning WORDPRESS_HOST or WORDPRESS_FORCE_TLS environment variables to the container. I can imagine a situation when someone decides to move their website from one domain to another or toggle HTTPS – the pod will need to be destroyed and recreated.

Indeed, the WORDPRESS_FORCE_TLS case wouldn't be an issue when switching domains, but toggling HTTPS would be. We'd need to document this.

> The host and the protocol do reach the container via HTTP headers, so in theory this information is enough. Because WORDPRESS_HOST is absent in deployment.yaml, I guess the domain detection already works well. In the ideal scenario, we just need to teach the bitnami wordpress image to detect the protocol as well and that will be it. After looking at the the current port, wordpress could check X-Forwarded-Proto and override the protocol if this value is present. Can there be a wordpress plugin for this that can be installed by default?

So my initial thought was that `$_SERVER['HTTPS']` would be empty because Ingress goes via HTTP, the `HTTP_HOST` header is fine, but we wouldn't be able to get the original scheme. As you say, I think we can solve this by checking `X-Forwarded-Proto` instead and using it to set the scheme.

**Comment by kachkaev** [comment]
How about adding these env variables to the image?

* `HTTP_HEADER_HOST`: undefined to take the value is taken from `$_SERVER['HTTP_HOST']`; can be set to `X-Forwarded-For` when running behind most of the reverse proxies
* `HTTP_HEADER_PROTOCOL`: undefined for a port-based detection, can be set to `X-Forwarded-Proto`
* `HTTP_HEADER_CLIENT_ADDRESS`: undefined for `$_SERVER['REMOTE_ADDR']`; can be set to `X-Forwarded-For`

These extra headers will make it possible to configure the image in pretty much any environment. If they are added, the chart will just need to get a new  `if` block in `deployment.yml` (if using ingress, assign three env variables to `X-...`). WDYT?

**Comment by kachkaev** [comment]
I was using a simple whoami image instead of wordpress to test which headers reach the pod: `gcr.io/google_containers/echoserver:1.8`. It listens on `8080` and prints all request details. Pretty handy!

**Comment by prydonius** [comment]
See #2484 for a solution to this, @kachkaev can you confirm if that works for you?

**Comment by kachkaev** [comment]
Thanks for sharing a link to #2484 @prydonius – I did not know about that PR. I haven't tried the new chart myself yet, but it seems like it might be a solution. There may be a problem with traefik not willing to talk to tls backends if the certs are not whitelisted, however I seem to be unable to test things right now.

If anyone else upgrades the wordpress chart to 0.7+ please share your experience to help decide if this issue can be now closed.

**Comment by prydonius** [comment]
cc @venezia @mattfarina

**Comment by kachkaev** [comment]
Unfortunately #2484 by @venezia did not help me. Here is what I did:

1. Generated a new dummy cert for wordpress:

    ```bash
    openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout /tmp/tls.key -out /tmp/tls.crt -subj "/CN=example.com"
    kubectl create secret tls wordpress.local-tls --key /tmp/tls.key --cert /tmp/tls.crt
    ```

1. Crafted my `values.yaml` so that it contains `tls: true`:

    ```yaml
    # ...
    ingress:
        enabled: true
        hosts:
        - name: example.com
          tls: true # ← this row
    # ...
    ```

1. Started a new wordpress instance and got `internal server error` when going to its url.

The problem ended up being in [traefik](https://github.com/containous/traefik), which I'm using as the ingress controller. It [does not support](https://github.com/containous/traefik/issues/364) skipping TLS verification for backends and so gives me this error when I'm trying to connect to wordpress through it:

```
Error forwarding to https://100.96.9.21:443, err: x509: cannot validate certificate for 100.96.9.21 because it doesn't contain any IP SANs
```

Kubernetes IP addresses are dynamic so I'm not sure I can ever generate a backend cert that will please traefik. So falling back to a [workaround](https://github.com/kubernetes/charts/issues/1567#issuecomment-319582737) by @keunwoo again for now.

Overall, I find it strange that I _must_ use TLS between the ingress controller and the wordpress pod. Other services like gitlab or nginx servers are able to figure out the true schema from the HTTP headers or env vars even when on port 80. It'd be great if this was possible for wordpress too, e.g. with the env vars I [proposed](https://github.com/kubernetes/charts/issues/1567#issuecomment-324394018) above.

**Comment by mattfarina** [comment]
@kachkaev A couple things...

1. Wordpress generates URLs to things like css files. The http(s) handling code would need to be changed there to help it better generate URLs for purposes like this.
2. Is there a reason the `InsecureSkipVerify` option in Traefik doesn't work? You can use it to skip verification of the cert for a backend. See https://docs.traefik.io/configuration/commons/. What am I missing? I'm not a regular user.

**Comment by gobengo** [comment]
It would be sufficient to get the following into the top of /opts/bitnami/wordpress/wp-config.php

```php
if ($_SERVER['FORCE_SSL']) {
  define('FORCE_SSL_ADMIN', true);
  define('FORCE_SSL_LOGIN', true);
}
if ($_SERVER['TRUST_HTTP_X_FORWARDED_PROTO']) {
  /* https://stackoverflow.com/a/41917594 */
  if (strpos($_SERVER['HTTP_X_FORWARDED_PROTO'], 'https') !== false)
         $_SERVER['HTTPS']='on';
}
```

**Comment by kachkaev** [comment]
Thanks for your thoughts @mattfarina. I was also thinking of telling traefik to ignore verify, but then decided to apply a [post-install hack](https://github.com/kubernetes/charts/issues/1567#issuecomment-319582737) instead.

The first problem was that `InsecureSkipVerify` could not be set just for wordpress, because there is [no ingress annotation](https://docs.traefik.io/configuration/backends/kubernetes/) for this. Doing so for all pods seems too dangerous, so I'm waiting for https://github.com/containous/traefik/issues/364 to be fixed.

Another less important issue is that the [traefik helm chart](https://github.com/kubernetes/charts/tree/master/stable/traefik) does not support `InsecureSkipVerify` but this is something I can theoretically fix myself.

I'm glad that @gobengo has suggested a solution for the bitnami image and I actually think that this is the right place for doing a fix. Apps like gitlab or sonatype nexus work well at HTTPS while the link between ingress and the pod is insecure — those pods determine the schema from headers and link to assets correctly. It would be great if wordpress could do that too, I'd just make headers configurable via env as [suggested above](https://github.com/kubernetes/charts/issues/1567#issuecomment-324394018). That's a few lines of php in any case.

**Comment by prydonius** [comment]
Thanks @kabakaev @gobengo - looks like they were going to make the change in https://github.com/bitnami/bitnami-docker-wordpress/issues/101 but there hasn't been an update there

**Comment by tompizmor** [comment]
Hi all! We just released WordPress 4.9.2-r1 docker image that should address this issue. By setting `SetEnvIf X-Forwarded-Proto https HTTPS=on` we let apache know we're behind an SSL reverse proxy and WordPress should use `https://` when forming the URLs.

**Comment by Fermiz** [comment]
Get the same issue, solved by adding `$_SERVER['HTTPS']='on';`  to `wp-config,php`, this make sure your php knows requests come from https.

**Comment by prydonius** [comment]
@tompizmor based on your earlier comment a manual change in wp-config.php shouldn't be required right?

**Comment by tompizmor** [comment]
You're right @prydonius. A manual change in the wp-config.php file should not be required.

@Fermiz, can you please check which docker image tag are you using for WordPress? According to my previous comment, it should be fixed since WordPress 4.9.2-r1.

**Comment by temujin9** [comment]
@tompizmor I'm encountering this on my GCE cluster. [My image](https://hub.docker.com/r/greenfieldguild/wordpress_cli/) is built off of `bitnami/wordpress:latest`. It adds, among other things, an edit to httpd.conf to set the one-liner you described. No such line is on your current latest image, and adding it fixes the problem.

If you let me know the correct file to edit, I'll make a PR to fix this.
